File: lib/src/main/java/io/github/goto1134/structurizr/export/d2/model/D2TextObject.java
package io.github.goto1134.structurizr.export.d2.model;

import com.structurizr.export.IndentingWriter;

import java.util.ArrayList;
import java.util.List;

public class D2TextObject extends D2Object {

    public static class Builder {
        private final String name;
        private final String language;
        private final String text;
        private final List<D2Property<D2Keyword, ?>> properties = new ArrayList<>();
        private final List<D2Property<D2StyleKeyword, ?>> style = new ArrayList<>();

        public Builder(String name, String language, String text) {
            this.name = name;
            this.language = language;
            this.text = text;
        }

        public Builder near(D2NearConstant near) {
            return near(near.toString());
        }

        public Builder near(String near) {
            properties.add(new D2Property<>(D2Keyword.NEAR, near));
            return this;
        }

        public Builder fontColor(String fontColor) {
            style.add(new D2WrappedStringProperty<>(D2StyleKeyword.TEXT_FONT_COLOR, fontColor));
            return this;
        }

        public Builder fontSize(Integer fontSize) {
            style.add(new D2Property<>(D2StyleKeyword.TEXT_FONT_SIZE, fontSize));
            return this;
        }

        public D2TextObject build() {
            return new D2TextObject(name, language, text, properties, style);
        }
    }

    public static Builder builder(String name, String language, String text) {
        return new Builder(name, language, text);
    }

    protected final String language;
    protected final String text;

    public D2TextObject(String name, String language, String text, List<D2Property<D2Keyword, ?>> properties, List<D2Property<D2StyleKeyword, ?>> style) {
        super(name, properties, style);
        this.language = language;
        this.text = text;
    }

    @Override
    public void writeHeader(IndentingWriter writer) {
        writer.writeLine(String.format("%s: |`%s", name, language));
        writer.indent();
        for (String line : text.split("\\R")) {
            writer.writeLine(line);
        }
        writer.outdent();
        writer.writeLine("`| {");
        writer.indent();
    }
}

File: lib/src/main/kotlin/io/github/goto1134/structurizr/export/d2/model/D2TextObject.kt
package io.github.goto1134.structurizr.export.d2.model

import com.structurizr.export.IndentingWriter
import io.github.goto1134.structurizr.export.d2.indented

class D2TextObject(
    name: String,
    private val language: String,
    private val text: String,
    properties: List<D2Property<D2Keyword, *>>,
    style: List<D2Property<D2StyleKeyword, *>>
) : D2Object(name, properties, style) {

    companion object {
        fun build(name: String, language: String, text: String, block: Builder.() -> Unit): D2TextObject {
            return Builder(name, language, text).apply(block).build()
        }
    }

    override fun openObject(writer: IndentingWriter) {
        writer.writeLine("$name: |`$language")
        writer.indented {
            text.splitToSequence("\\R".toRegex()).forEach {
                writeLine(it)
            }
        }
        writer.writeLine("`| {")
    }

    class Builder(private val name: String, private val language: String, private val text: String) {
        private val properties: MutableList<D2Property<D2Keyword, *>> = ArrayList()
        private val style: MutableList<D2Property<D2StyleKeyword, *>> = ArrayList()
        fun near(near: D2NearConstant): Builder = apply {
            properties.add(D2Property(D2Keyword.NEAR, near.toString()))
        }

        fun fontColor(fontColor: String?): Builder = apply {
            if (fontColor != null) style.add(D2WrappedStringProperty(D2StyleKeyword.TEXT_FONT_COLOR, fontColor))
        }

        fun fontSize(fontSize: Int): Builder = apply {
            style.add(D2Property(D2StyleKeyword.TEXT_FONT_SIZE, fontSize))
        }

        fun build(): D2TextObject {
            return D2TextObject(name, language, text, properties, style)
        }
    }
}
